HPCC-20035 The remoteread.ecl doesn't compatible with parallel query execution mode

AttilaVamos opened the pull request:
Add WUID dependent suffix for filenames to generate unique logical
filenames pro execution.

Remove logical filenames from dataset, based on the dinamic filenames
can't compare with static key file.

Add cleanup to remove generated logical files.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually: Execute remoteread.ecl in a loop when parallel query
execution is enabled.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a60abaedd6ec45b6a3bd3fda50c5c5fc0f3f88e7
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  172 sec (00:02:52) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1272 sec (00:21:12) |   17 sec (00:00:17) | 1574 sec (00:26:14) |

AttilaVamos commented:
@richardkchapman Please review. I removed the logical file names from data sets. Can it cause any problem?

richardkchapman commented:
The problem is that this test is designed to test that the virtual filename is returned properly (among other things).

You could add a unique suffix and remove that in the transform before outputting the result.

richardkchapman commented:
@AttilaVamos

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9862be604235d1761a61e38efa4e9ec0e1713136
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  170 sec (00:02:50) |   62 sec (00:01:02) |    6 sec (00:00:06) |   27 sec (00:00:27) | 1272 sec (00:21:12) |   18 sec (00:00:18) | 1572 sec (00:26:12) |

AttilaVamos commented:
@richardkchapman Reworked. Unique logical filename trimming ,storing and checking. Please review.

richardkchapman commented on testing/regress/ecl/remoteread.ecl:
```diff
@@ -59,9 +61,13 @@ rec_vf_trans := RECORD
-ds_comp := DATASET(fname_comp, rec_vf, FLAT);
-ds_trans := DATASET(fname, rec_vf_trans, FLAT);
+string trimmedLogicalFilename(STRING filename) := FUNCTION
+    return filename[1..(std.str.find(std.str.ToLowerCase(trim(filename)), '-w2')-1)];
```
This will fail in 1000 years...

richardkchapman commented:
@AttilaVamos Please squash

AttilaVamos committed:
HPCC-20035 The remoteread.ecl doesn't compatible with parallel query execution mode

Add WUID dependent suffix for filenames to generate unique logical
filenames pro execution.

Remove logical filenames from dataset, based on the dinamic filenames
can't compare with static key file.

Add cleanup to remove generated logical files.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fc62ebef41ebfd73750482fcbfbb47bf17ca0092
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  169 sec (00:02:49) |   62 sec (00:01:02) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1283 sec (00:21:23) |   17 sec (00:00:17) | 1579 sec (00:26:19) |

AttilaVamos commented:
@richardkchapman The "This will fail in 1000 years..." problem fixed and commits are squashed.
